File: frontend/tests/security/corpus/hostileEpub.ts
/**
 * Hostile EPUB corpus builders (issue #87, E-1..E-5). Runtime-generated,
 * never committed as binaries; deterministic so a failing case reproduces
 * byte for byte. The vitest corpus tests consume these directly; the E2E
 * security spec keeps its own copy because e2e is a separate package and
 * cannot import from the frontend test tree.
 */

const CRC_TABLE = (() => {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n += 1) {
    let c = n;
    for (let k = 0; k < 8; k += 1) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    table[n] = c >>> 0;
  }
  return table;
})();

function crc32(data: Buffer): number {
  let crc = 0xffffffff;
  for (const byte of data) {
    crc = (CRC_TABLE[(crc ^ byte) & 0xff] ?? 0) ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

function u16(value: number): Buffer {
  return Buffer.from([value & 0xff, (value >>> 8) & 0xff]);
}

function u32(value: number): Buffer {
  return Buffer.from([
    value & 0xff,
    (value >>> 8) & 0xff,
    (value >>> 16) & 0xff,
    (value >>> 24) & 0xff,
  ]);
}

function buildZip(entries: { name: string; data: string }[]): Buffer {
  const locals: Buffer[] = [];
  const central: Buffer[] = [];
  let offset = 0;
  for (const { name, data } of entries) {
    const bytes = Buffer.from(data, "utf8");
    const nameBytes = Buffer.from(name, "utf8");
    const crc = crc32(bytes);
    const local = Buffer.concat([
      u32(0x04034b50),
      u16(20),
      u16(0),
      u16(0),
      u16(0),
      u16(0),
      u32(crc),
      u32(bytes.length),
      u32(bytes.length),
      u16(nameBytes.length),
      u16(0),
      nameBytes,
      bytes,
    ]);
    const directory = Buffer.concat([
      u32(0x02014b50),
      u16(20),
      u16(20),
      u16(0),
      u16(0),
      u16(0),
      u16(0),
      u32(crc),
      u32(bytes.length),
      u32(bytes.length),
      u16(nameBytes.length),
      u16(0),
      u16(0),
      u16(0),
      u16(0),
      u32(0),
      u32(offset),
      nameBytes,
    ]);
    locals.push(local);
    central.push(directory);
    offset += local.length;
  }
  const centralBytes = Buffer.concat(central);
  const end = Buffer.concat([
    u32(0x06054b50),
    u16(0),
    u16(0),
    u16(entries.length),
    u16(entries.length),
    u32(centralBytes.length),
    u32(offset),
    u16(0),
  ]);
  return Buffer.concat([...locals, centralBytes, end]);
}

export interface HostileEpub {
  name: string;
  invariant: string;
  /** The chapter bytes the policy layer must fence. */
  chapter: string;
  bytes: Buffer;
}

const XHTML_PROLOG = `<?xml version="1.0" encoding="UTF-8"?>
<html xmlns="http://www.w3.org/1999/xhtml"><head><title>c</title></head><body>`;

function xhtml(body: string): string {
  return `${XHTML_PROLOG}${body}</body></html>`;
}

function epub(title: string, chapter: string): Buffer {
  const opf = `<?xml version="1.0" encoding="UTF-8"?>
<package xmlns="http://www.idpf.org/2007/opf" version="3.0" unique-identifier="id">
  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
    <dc:identifier id="id">urn:uuid:${title.toLowerCase().replace(/\s+/g, "-")}</dc:identifier>
    <dc:title>${title}</dc:title>
    <dc:language>en</dc:language>
  </metadata>
  <manifest>
    <item id="c1" href="chapter1.xhtml" media-type="application/xhtml+xml"/>
  </manifest>
  <spine><itemref idref="c1"/></spine>
</package>`;
  const container = `<container xmlns="urn:oasis:names:tc:opendocument:xmlns:container"><rootfiles><rootfile full-path="content.opf"/></rootfiles></container>`;
  return buildZip([
    { name: "mimetype", data: "application/epub+zip" },
    { name: "META-INF/container.xml", data: container },
    { name: "content.opf", data: opf },
    { name: "chapter1.xhtml", data: chapter },
  ]);
}

export const HOSTILE_EPUB_CORPUS: readonly HostileEpub[] = [
  {
    name: "hostile-scripted",
    invariant: "E-1",
    chapter: xhtml(`<p>text</p><script>alert(1)</script>`),
    bytes: epub("Hostile Scripted Book", xhtml(`<p>text</p><script>alert(1)</script>`)),
  },
  {
    name: "hostile-active",
    invariant: "E-2",
    chapter: xhtml(
      `<p>text</p><iframe src="page.xhtml"></iframe>` +
        `<object data="evil.swf" type="application/x-shockwave-flash"></object>` +
        `<embed src="evil.swf" type="application/x-shockwave-flash">`,
    ),
    bytes: epub(
      "Hostile Active Book",
      xhtml(
        `<p>text</p><iframe src="page.xhtml"></iframe>` +
          `<object data="evil.swf" type="application/x-shockwave-flash"></object>` +
          `<embed src="evil.swf" type="application/x-shockwave-flash">`,
      ),
    ),
  },
  {
    name: "hostile-external",
    invariant: "E-3",
    chapter: xhtml(
      `<p>text</p><img src="https://evil.example/track.png" alt="beacon" />` +
        `<link rel="stylesheet" href="https://evil.example/style.css" />`,
    ),
    bytes: epub(
      "Hostile External Book",
      xhtml(
        `<p>text</p><img src="https://evil.example/track.png" alt="beacon" />` +
          `<link rel="stylesheet" href="https://evil.example/style.css" />`,
      ),
    ),
  },
  {
    name: "hostile-path-probe",
    invariant: "E-4",
    chapter: xhtml(
      `<p>text</p><a href="file:///home/user/.ssh/id_rsa">probe</a>` +
        `<a href="tuxbooks://book/999/chapter1.xhtml">cross-book</a>`,
    ),
    bytes: epub(
      "Hostile Path Probe Book",
      xhtml(
        `<p>text</p><a href="file:///home/user/.ssh/id_rsa">probe</a>` +
          `<a href="tuxbooks://book/999/chapter1.xhtml">cross-book</a>`,
      ),
    ),
  },
  {
    name: "hostile-encoding",
    invariant: "E-1",
    chapter: `<?xml version="1.0" encoding="utf-7"?>
<html xmlns="http://www.w3.org/1999/xhtml"><head><title>c</title></head><body><p>text</p><div onload="alert(1)">x</div></body></html>`,
    bytes: epub(
      "Hostile Encoding Book",
      `<?xml version="1.0" encoding="utf-7"?>
<html xmlns="http://www.w3.org/1999/xhtml"><head><title>c</title></head><body><p>text</p><div onload="alert(1)">x</div></body></html>`,
    ),
  },
];

File: sidecar/tests/fixtures/security/hostile_epub.rs
//! Hostile EPUB corpus builders (issue #87, E/R invariants).
//!
//! Repo convention: hostile fixtures are generated at runtime, never
//! committed as binaries. This module is the shared loader for those
//! builders — included by `tests/security_corpus.rs` via `#[path]` and
//! reusable as fuzz seeds by #88. Every builder is deterministic (no
//! clock, fixed seeds) so a failing case reproduces byte for byte.

use std::io::Write as _;
use std::path::Path;

pub(crate) const MIMETYPE: &str = "application/epub+zip";

pub(crate) fn container_xml() -> String {
    r#"<container xmlns="urn:oasis:names:tc:opendocument:xmlns:container"><rootfiles><rootfile full-path="content.opf"/></rootfiles></container>"#
        .to_string()
}

pub(crate) fn opf(manifest_extra: &str, cover: bool) -> String {
    let cover_item = if cover {
        r#"<item id="cover" href="cover.png" media-type="image/png" properties="cover-image"/>"#
    } else {
        ""
    };
    format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<package xmlns="http://www.idpf.org/2007/opf" version="3.0" unique-identifier="id">
  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
    <dc:identifier id="id">urn:uuid:corpus</dc:identifier>
    <dc:title>Hostile Corpus Book</dc:title>
    <dc:language>en</dc:language>
  </metadata>
  <manifest>
    <item id="c1" href="chapter1.xhtml" media-type="application/xhtml+xml"/>
    {cover_item}
    {manifest_extra}
  </manifest>
  <spine><itemref idref="c1"/></spine>
</package>"#
    )
}

pub(crate) const CHAPTER: &str = r#"<?xml version="1.0" encoding="UTF-8"?>
<html xmlns="http://www.w3.org/1999/xhtml"><head><title>c</title></head><body><p>chapter text</p></body></html>"#;

/// Write a ZIP with deflate members, like the parser's own test helper.
pub(crate) fn write_zip(path: &Path, entries: &[(String, Vec<u8>)]) {
    let file = std::fs::File::create(path).unwrap();
    let mut zip = zip::ZipWriter::new(file);
    for (name, data) in entries {
        zip.start_file(name.clone(), zip::write::SimpleFileOptions::default())
            .unwrap();
        zip.write_all(data).unwrap();
    }
    zip.finish().unwrap();
}

/// Deterministic pseudo-random byte stream (xorshift64): hostile members
/// must deflate badly (incompressible) without a dependency or clock.
pub(crate) struct Noise(u64);

impl Noise {
    pub(crate) fn new(seed: u64) -> Self {
        Self(seed | 1)
    }

    pub(crate) fn next_byte(&mut self) -> u8 {
        self.0 ^= self.0 << 13;
        self.0 ^= self.0 >> 7;
        self.0 ^= self.0 << 17;
        (self.0 >> 24) as u8
    }

    pub(crate) fn bytes(&mut self, len: usize) -> Vec<u8> {
        (0..len).map(|_| self.next_byte()).collect()
    }
}

pub(crate) fn epub_bytes(opf_xml: &str, chapter: &str) -> Vec<(String, Vec<u8>)> {
    vec![
        ("mimetype".to_string(), MIMETYPE.as_bytes().to_vec()),
        (
            "META-INF/container.xml".to_string(),
            container_xml().into_bytes(),
        ),
        ("content.opf".to_string(), opf_xml.as_bytes().to_vec()),
        ("chapter1.xhtml".to_string(), chapter.as_bytes().to_vec()),
    ]
}

/// E-1 fixture: a manifest item declaring a script media type (not in the
/// spine, so it ships a script resource without a scripted spine document).
pub(crate) fn scripted_manifest_item(path: &Path) {
    write_zip(
        path,
        &epub_bytes(
            &opf(
                r#"<item id="evil" href="evil.js" media-type="text/javascript"/>"#,
                false,
            ),
            CHAPTER,
        ),
    );
}

/// R-1 fixture: a cover member of incompressible noise, so its compressed
/// size far exceeds a small `max_compressed_member_bytes` while every other
/// quota passes.
pub(crate) fn incompressible_cover(path: &Path) {
    let noise = Noise::new(0x5EED_0001).bytes(4 << 10);
    let mut entries = epub_bytes(&opf("", true), CHAPTER);
    entries.push(("cover.png".to_string(), noise));
    write_zip(path, &entries);
}

/// R-1/E-5 fixture: members nested 200 directories deep — legal shapes that
/// must stay bounded, plus an over-long member name that must miss cleanly.
pub(crate) fn deep_paths(path: &Path) {
    let deep_dir = "d/".repeat(200);
    let mut entries = epub_bytes(&opf("", false), CHAPTER);
    entries.push((format!("{deep_dir}leaf.xhtml"), CHAPTER.as_bytes().to_vec()));
    write_zip(path, &entries);
}

pub(crate) fn deep_member_name() -> String {
    format!("{}leaf.xhtml", "d/".repeat(200))
}

/// R-1 fixture: a truncated archive (valid EPUB bytes cut mid central
/// directory).
pub(crate) fn truncated(path: &Path) {
    let mut entries = epub_bytes(&opf("", false), CHAPTER);
    entries.push(("filler.bin".to_string(), Noise::new(7).bytes(64)));
    let scratch = path.with_extension("full");
    write_zip(&scratch, &entries);
    let bytes = std::fs::read(&scratch).unwrap();
    let cut = bytes.len() * 3 / 5;
    std::fs::write(path, &bytes[..cut]).unwrap();
    let _ = std::fs::remove_file(&scratch);
}

/// R-1 fixture: a valid EPUB whose central-directory signature is corrupted.
pub(crate) fn corrupt_central_directory(path: &Path) {
    let mut entries = epub_bytes(&opf("", false), CHAPTER);
    entries.push(("filler.bin".to_string(), Noise::new(11).bytes(64)));
    let scratch = path.with_extension("full");
    write_zip(&scratch, &entries);
    let mut bytes = std::fs::read(&scratch).unwrap();
    let sig = [0x50u8, 0x4b, 0x01, 0x02];
    let pos = bytes
        .windows(4)
        .rposition(|w| w == sig)
        .expect("fixture must contain a central-directory signature");
    bytes[pos] = 0x51;
    std::fs::write(path, &bytes).unwrap();
    let _ = std::fs::remove_file(&scratch);
}

/// R-1 fixture: OPF carrying a billion-laughs style internal entity
/// expansion aimed at the title string.
pub(crate) fn entity_bomb(path: &Path) {
    let mut entities = String::from("<!ENTITY e0 \"AAAAAAAABBBBBBBBCCCCCCCCDDDDDDDD\">");
    for level in 1..8 {
        let refs = format!("&e{};", level - 1);
        entities.push_str(&format!("<!ENTITY e{level} \"{}\">", refs.repeat(10)));
    }
    let opf_xml = format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE package [{}]>
<package xmlns="http://www.idpf.org/2007/opf" version="3.0" unique-identifier="id">
  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
    <dc:identifier id="id">urn:uuid:corpus</dc:identifier>
    <dc:title>&e7;</dc:title>
    <dc:language>en</dc:language>
  </metadata>
  <manifest>
    <item id="c1" href="chapter1.xhtml" media-type="application/xhtml+xml"/>
  </manifest>
  <spine><itemref idref="c1"/></spine>
</package>"#,
        entities
    );
    write_zip(path, &epub_bytes(&opf_xml, CHAPTER));
}

/// E/R fixture: malformed cover image and font members (garbage bytes) that
/// must travel through the pipeline inert — extracted and served
/// byte-identically, never decoded or rewritten.
pub(crate) fn malformed_cover_and_font(path: &Path) -> (Vec<u8>, Vec<u8>) {
    let cover = Noise::new(0xBEEF).bytes(2 << 10);
    let font = Noise::new(0xF0AD).bytes(1 << 10);
    let opf_xml = opf(
        r#"<item id="font" href="evil.woff" media-type="font/woff"/>"#,
        true,
    );
    let mut entries = epub_bytes(&opf_xml, CHAPTER);
    entries.push(("cover.png".to_string(), cover.clone()));
    entries.push(("evil.woff".to_string(), font.clone()));
    write_zip(path, &entries);
    (cover, font)
}
